lib/src/r_tree/quickselect.dart

```diff
@@ -0,0 +1,114 @@
+part of r_tree;
+// Port of https://github.com/mourner/quickselect.
+
+// sort an array so that items come in groups of n unsorted items, with groups sorted between each other;
+// combines selection algorithm with binary divide & conquer approach
+multiSelect<E>(List<E> arr, int left, int right, int n, num Function(E) getter) {
+  final stack = [left, right];
+  final compare = (E a, E b) => getter(a).compareTo(getter(b));
+
+  while (stack.isNotEmpty) {
+    right = stack.removeLast();
+    left = stack.removeLast();
+
+    if (right - left <= n) {
+      continue;
+    }
+
+    final mid = left + ((right - left).toDouble() / n / 2).ceil() * n;
+    quickSelect(arr, mid, left, right, compare);
+    stack.addAll([left, mid, mid, right]);
+  }
+}
+
+/// This function implements a fast
+/// [selection algorithm](https://en.wikipedia.org/wiki/Selection_algorithm)
+/// (specifically, [Floyd-Rivest selection](https://en.wikipedia.org/wiki/Floyd%E2%80%93Rivest_algorithm)).
+///
+/// Rearranges items so that all items in the `[left, k]` are the smallest.
+/// The `k`-th element will have the `(k - left + 1)`-th smallest value in `[left, right]`.
+///
+/// - [arr]: the list to partially sort (in place)
+/// - [k]: middle index for partial sorting (as defined above)
+/// - [left]: left index of the range to sort (`0` by default)
+/// - [right]: right index (last index of the array by default)
```

Comment:
Maybe I'm code-review blind right now, where is `right` defaulting to the last index of the array?

Reply:
whoops, I changed it as part of the sound null safety and didn't think to update the docstring